crates/accelerate/src/equivalence.rs

```diff
@@ -0,0 +1,786 @@
+        slf.graph = GraphType::new();
+        for node_weight in graph_nodes {
+            slf.graph.add_node(node_weight.extract()?);
+        }
+        for edge in graph_edges {
+            let (source_node, target_node, edge_weight) = edge.extract()?;
+            slf.graph.add_edge(
+                NodeIndex::new(source_node),
+                NodeIndex::new(target_node),
+                edge_weight,
+            );
+        }
+        slf.key_to_node_index = state
+            .get_item("key_to_node_index")?
+            .unwrap()
+            .extract::<IndexMap<Key, usize>>()?
+            .into_iter()
+            .map(|(key, val)| (key, NodeIndex::new(val)))
+            .collect();
+        slf._graph = None;
+        Ok(())
+    }
+}
+
+// Rust native methods
+impl EquivalenceLibrary {
+    fn add_equivalence(
+        &mut self,
+        py: Python,
+        gate: &GateOper,
+        equivalent_circuit: CircuitRep,
+    ) -> PyResult<()> {
+        raise_if_shape_mismatch(gate, &equivalent_circuit)?;
+        raise_if_param_mismatch(
+            py,
+            &gate.params,
+            equivalent_circuit.0.unsorted_parameters(py)?,
+        )?;
+        let op_ref = gate.operation.view();
+        let key: Key = Key {
+            name: op_ref.name().to_string(),
+            num_qubits: op_ref.num_qubits(),
+        };
+        let equiv = Equivalence {
+            params: gate.params.clone(),
+            circuit: equivalent_circuit.clone(),
+        };
+
+        let target = self.set_default_node(key);
+        if let Some(node) = self.graph.node_weight_mut(target) {
+            node.equivs.push(equiv.clone());
+        }
+        let sources: IndexSet<Key, RandomState> =
+            IndexSet::from_iter(equivalent_circuit.0.iter().map(|inst| Key {
+                name: inst.op.view().name().to_string(),
+                num_qubits: inst.op.view().num_qubits(),
+            }));
+        let edges = Vec::from_iter(sources.iter().map(|source| {
+            (
+                self.set_default_node(source.clone()),
+                target,
+                EdgeData {
+                    index: self.rule_id,
+                    num_gates: sources.len(),
+                    rule: equiv.clone(),
+                    source: source.clone(),
+                },
+            )
+        }));
+        for edge in edges {
+            self.graph.add_edge(edge.0, edge.1, Some(edge.2));
+        }
+        self.rule_id += 1;
+        self._graph = None;
+        Ok(())
+    }
+
+    /// Rust native equivalent to `EquivalenceLibrary.has_entry()`
+    ///
+    /// Check if a library contains any decompositions for gate.
+    ///
+    /// # Arguments:
+    /// * `operation` OperationType: A Gate instance.
+    ///
+    /// # Returns:
+    /// `bool`: `true` if gate has a known decomposition in the library.
+    ///         `false` otherwise.
+    pub fn has_entry(&self, operation: &PackedOperation) -> bool {
```

Comment:
I think this does not need to be `pub`.  Same thing occurs in other places as well.

Reply:
Hmm... Why is that the case here? I understand that some methods are not currently being used, however this is part of the `EquivalenceLibrary`'s original functionality and it was a public function by definition.